tools/profiling/dbpreader.c

```diff
@@ -378,55 +417,188 @@ void dbp_iterator_delete(dbp_event_iterator_t *it)
     free(it);
 }
 
+static inline int dbp_events_match(const dbp_event_t *s, const dbp_event_t *e)
+{
+    int s_key = dbp_event_get_key(s);
+    int e_key = dbp_event_get_key(e  );
+    return ( (KEY_IS_START(s_key)          && KEY_IS_END(e_key))            &&
+             (BASE_KEY(    s_key)          == BASE_KEY(  e_key))            &&
+             (dbp_event_get_event_id(   s) == dbp_event_get_event_id(   e)) &&
+             (dbp_event_get_taskpool_id(s) == dbp_event_get_taskpool_id(e)) &&
+             (dbp_event_get_timestamp(  s) <= dbp_event_get_timestamp(  e)) );
+}
+
+/* minimum allocation count for cache */
+#define EVENT_CACHE_MIN_ALLOC 64
+/* build a "cache" of events where the end event
+ * does not immediately follow the start event */
+static void build_unmatched_events_in_thread(dbp_thread_t *thr)
+{
+    const dbp_event_t      *e1,  *e2;
+    dbp_event_iterator_t   *i1,  *i2;
+    int key2;
+
+    event_cache_key_t  *cache_key;
+    event_cache_item_t *cache_item;
+    size_t              cache_index;
+
+    /* lock cache; we're modifying volatile state! */
+    pthread_mutex_lock(&thr->cache.mtx);
+    if( thr->cache.done ) {
+        /* cache already built, we don't need to do anything */
+        goto build_events_done;
+    }
+
+    /* iterator 1 points to current event */
+    i1 = dbp_iterator_new_from_thread( thr );
+    e1 = dbp_iterator_current( i1 );
+
+    /* iterator 2 points to next event */
+    i2 = dbp_iterator_new_from_thread( thr );
+    e2 = dbp_iterator_next( i2 );
+
+    while( NULL != e2 ) {
+        key2 = dbp_event_get_key(e2);
+
+        /* if e2 is end event, but e1 doesn't match, then e2 out of order */
```

Comment:
'out of order' is confusing in the comment here. I would replace with 'then store e2 position in the cache'.

Reply:
That makes sense, I'll do that in a bit once I have access to the machine I'm running on again.